### PR TITLE
Allow plugin download to use an insecure repository

### PR DESCRIPTION
**Pull request by akram**

### CODE REVIEW

**Comment by akram** [comment]
hi @adambkaplan , can you PTAL ?

**Comment by adambkaplan** [comment]
@akram I assume this is to address https://jira.coreos.com/browse/DEVEXP-367 and https://bugzilla.redhat.com/show_bug.cgi?id=1693533 ?

**Comment by gabemontero** [comment]
@akram @adambkaplan I'm taking a look

**Comment by gabemontero** [comment]
Also, to help @akram on the e2e-aws-jenkins failure, I took a peek for him.  It failed in the client plugin test becase of an apparent expired certificate.  From https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_jenkins/866/pull-ci-openshift-jenkins-master-e2e-aws-jenkins/288:

```
[Pipeline] End of Pipeline
hudson.AbortException: rsh returned an error;
{reference={}, err=Error from server: error dialing backend: x509: certificate signed by unknown authority
, verb=rsh, cmd=oc --server=https://172.30.0.1:443 --certificate-authority=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt --namespace=e2e-test-jenkins-pipeline-xskb5 --token=XXXXX rsh mongodb-1-4qrs7 ps ax , out=, status=1}

	at com.openshift.jenkins.plugins.OpenShiftDSL$Result.failIf(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:1030)
	at com.openshift.jenkins.plugins.OpenShiftDSL.simplePassthrough(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:980)
	at com.openshift.jenkins.plugins.OpenShiftDSL.rsh(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:985)
	at WorkflowScript.run(WorkflowScript:185)
	at com.openshift.jenkins.plugins.OpenShiftDSL$OpenShiftResourceSelector.untilEach(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:1353)
	at com.openshift.jenkins.plugins.OpenShiftDSL$OpenShiftResourceSelector.withEach(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:1629)
	at com.openshift.jenkins.plugins.OpenShiftDSL$OpenShiftResourceSelector.untilEach(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:1352)
	at com.openshift.jenkins.plugins.OpenShiftDSL$OpenShiftResourceSelector.watch(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:1321)
	at ___cps.transform___(Native Method)

```

On the surface that feels like a api.ci / openshift CICD flake.  I saw similar ones in unrelated networking e2es in the e2e-aws run:

`        StdErr: "Error from server: error dialing backend: x509: certificate signed by unknown authority",`

**Comment by akram** [comment]
> https://bugzilla.redhat.com/show_bug.cgi?id=1693533



> @akram I assume this is to address https://jira.coreos.com/browse/DEVEXP-367 and https://bugzilla.redhat.com/show_bug.cgi?id=1693533 ?

Hi @adambkaplan 
I didn't know about the BZ and the JIRA.  My initial PR was done maybe in last December and the BZ has been reported on march.
Reading the BZ, it looks like they are asking to expose CURL_OPTIONS which may have more impacts and would require more input filtering.

**Comment by akram** [comment]
> @akram since this is a RH subscription customer support case calling for this we should expose this tuning option:
> 
> 1. update the README's env var section to explain it
> 2. update the templates under https://github.com/openshift/jenkins/tree/master/openshift/templates
> 3. update the docs under https://github.com/openshift/openshift-docs ... as those are still in a bit of flux, reach out to @bmcelvee to nail down the right branch/location

Hi @gabemontero , thank you for reviewing.
Definitely, I was thinking about updating the README. And OK, I will update the template also. I will do it in separate commits and squash when all validated.

File: README.md [hunk 1]
```diff
@@ -148,6 +148,7 @@ initialization by passing `-e VAR=VALUE` to the Docker run command.
...
 |  `MAVEN_SLAVE_IMAGE`   | Setting this value overrides the image used for the default maven agent pod configuration.  For 3.x, the default maven agent pod uses `docker.io/openshift/jenkins-agent-maven-35-centos7` or `registry.redhat.io/openshift3/jenkins-agent-maven-35-rhel7` depending whether you are running the centos or rhel version of the Jenkins image.  For 4.x, the image is included in the 4.0 payload via an imagestream in the openshift namespace, and the image spec points to the internal image registry.  If you are running this image outside of OpenShift, you must either set this environment variable or manually update the setting to an accessible image spec. This variable must be set before Jenkins starts the first time for it to have an effect. |
+|  `JENKINS_UC_INSECURE`       | Set this variable to 'true' or '1', if your Jenkins Update Center repository uses a self-signed certificate with an unknown Certificate Authority. This could be useful if you are using an internal repository to store additional plugins. default: 'false' |
```
Comment (gabemontero) [inline README.md]: clarify that this is related to the install plugins invocation that occurs when a) building the jenkins image, b) building extensions to the jenkins image, or c) when you run the jenkins image and leverage one of our options for installing additional plugins

File: 2/contrib/jenkins/install-plugins.sh [hunk 2]
```diff
@@ -225,8 +225,14 @@ function doDownload() {
         url="$JENKINS_UC_DOWNLOAD/plugins/$plugin/$version/${plugin}.hpi"
     fi
 
+    JENKINS_UC_INSECURE=${JENKINS_UC_INSECURE:-"false"}
+    if [[ -n "$JENKINS_UC_INSECURE" ]] && [[ "$JENKINS_UC_INSECURE" = true || "$JENKINS_UC_INSECURE" = 1 ]]; then
```
Comment (gabemontero) [inline 2/contrib/jenkins/install-plugins.sh]: why support both true and 1 ... with other env vars, we employed a patter that if
- the env was not empty
- and it was set to something other than false
toggle the behavior

Comment (akram) [reply 2/contrib/jenkins/install-plugins.sh]: Resolved. I made the test syntax a bit verbose to protect unset variables checking in shell, but it now uses the logic you described.

File: README.md (hunk 1 above)
Comment (akram) [reply README.md]: Should be fixed now to reflect these comments.

**Comment by akram** [comment]
> @akram since this is a RH subscription customer support case calling for this we should expose this tuning option:
> 
> 1. update the README's env var section to explain it
> 2. update the templates under https://github.com/openshift/jenkins/tree/master/openshift/templates
> 3. update the docs under https://github.com/openshift/openshift-docs ... as those are still in a bit of flux, reach out to @bmcelvee to nail down the right branch/location

Documentation update is in this branch:
https://github.com/akram/openshift-docs/tree/allow-plugin-download-to-use-an-insecure-repository which I based on enterprise-3.11 .


I contacted @bmcelvee  to know which branch to rely on to create the PR. I am waiting for her reply.

**Comment by akram** [comment]
PRs for documentation:
https://github.com/openshift/openshift-docs/pull/15258
https://github.com/openshift/openshift-docs/pull/14227
https://github.com/bmcelvee/openshift-docs/pull/2

File: README.md [hunk 3]
```diff
@@ -148,6 +148,7 @@ initialization by passing `-e VAR=VALUE` to the Docker run command.
...
+|  `JENKINS_UC_INSECURE`       | When your Jenkins Update Center repository is using a self-signed certificate with an unknown Certificate Authority, this variable allows to bypass the repository's SSL certificate check. The variable applies to plugins downloads which may occur during Jenkins image build or if you build an extension of the jenkins image or if you run the jenkins image and leverage one of the options to download additional plugins (use of s2i whith plugins.txt or use of `INSTALL_PLUGINS` environment variable. |
```
Comment (gabemontero) [inline README.md]: A couple of minor grammatical, stylistic, items:
- `this variable allows to bypass` -> `this variable allows one to bypass`
- `applies to plugins downloads` -> `applies to download of the plugins`
- put commas i.e. `,` before each `or`
- change any `jenkins` to `Jenkins`
- you need an ending `)`

File: openshift/templates/jenkins-ephemeral.json [hunk 4]
```diff
@@ -292,6 +296,12 @@
       "displayName": "Jenkins ImageStreamTag",
       "description": "Name of the ImageStreamTag to be used for the Jenkins image.",
       "value": "jenkins:2"
+    },
+    {
+      "name": "JENKINS_UC_INSECURE",
+      "displayName": "Allows use of Jenkins Update Center repository with invalid SSL certificate",
+      "description": "Whether to allow use of a Jenkins Update Center that uses invalid certificate (self-signed, unknown CA). If any value other than 'false', certificate check is bypassed.",
```
Comment (gabemontero) [inline openshift/templates/jenkins-ephemeral.json]: Add a note that the default is to enforce the certificate check

File: openshift/templates/jenkins-persistent.json [hunk 5]
```diff
@@ -326,6 +330,12 @@
       "displayName": "Fatal Error Log File",
       "description": "When a fatal error occurs, an error log is created with information and the state obtained at the time of the fatal error.",
       "value": "false"
+    },
+    {
+      "name": "JENKINS_UC_INSECURE",
+      "displayName": "Allows use of Jenkins Update Center repository with invalid SSL certificate",
+      "description": "Whether to allow use of a Jenkins Update Center that uses invalid certificate (self-signed, unknown CA). If any value other than 'false', certificate check is bypassed.",
```
Comment (gabemontero) [inline openshift/templates/jenkins-persistent.json]: Add a note that the default is to enforce the certificate check

File: README.md (hunk 3 above)
Comment (akram) [reply README.md]: Fixed

File: openshift/templates/jenkins-persistent.json (hunk 5 above)
Comment (akram) [reply openshift/templates/jenkins-persistent.json]: Fixed

File: openshift/templates/jenkins-ephemeral.json (hunk 4 above)
Comment (akram) [reply openshift/templates/jenkins-ephemeral.json]: Fixed

**Pull request merged**